libcaf_net/test/transport_worker.cpp

```diff
@@ -0,0 +1,222 @@
+    transport.write_packet(span<byte>{}, msg->payload);
+  }
+
+  template <class Parent>
+  void handle_data(Parent&, span<const byte> data) {
+    auto& buf = res_->data_buffer;
+    buf.clear();
+    buf.insert(buf.begin(), data.begin(), data.end());
+  }
+
+  template <class Parent>
+  void resolve(Parent&, const std::string& path, actor listener) {
+    res_->resolve_path = path;
+    res_->resolve_listener = std::move(listener);
+  }
+
+  template <class Parent>
+  void timeout(Parent&, atom_value value, uint64_t id) {
+    res_->timeout_value = value;
+    res_->timeout_id = id;
+  }
+
+  void handle_error(sec err) {
+    res_->err = err;
+  }
+
+  static expected<std::vector<byte>> serialize(actor_system& sys,
+                                               const type_erased_tuple& x) {
+    std::vector<byte> result;
+    serializer_impl<std::vector<byte>> sink{sys, result};
+    if (auto err = message::save(sink, x))
+      return err;
+    return result;
+  }
+
+private:
+  std::shared_ptr<application_result> res_;
+};
+
+class dummy_transport {
+public:
+  using transport_type = dummy_transport;
+
+  using application_type = dummy_application;
+
+  dummy_transport(std::shared_ptr<transport_result> res) : res_(res) {
+    // nop
+  }
+
+  template <class Header>
+  void write_packet(Header header, span<const byte> payload, ip_endpoint ep) {
+    auto& buf = res_->packet_buffer;
+    buf.insert(buf.begin(), header.begin(), header.end());
+    buf.insert(buf.begin(), payload.begin(), payload.end());
+    res_->ep = ep;
+  }
+
+private:
+  std::shared_ptr<transport_result> res_;
+};
+
+struct fixture : test_coordinator_fixture<>, host_fixture {
+  using worker_type = transport_worker<dummy_application, ip_endpoint>;
+
+  fixture()
+    : transport_results{std::make_shared<transport_result>()},
+      application_results{std::make_shared<application_result>()},
+      transport(transport_results),
+      worker{dummy_application{application_results}} {
+    mpx = std::make_shared<multiplexer>();
+    if (auto err = mpx->init())
+      CAF_FAIL("mpx->init failed: " << sys.render(err));
+    if (auto err = parse("[::1]:12345", ep))
+      CAF_FAIL("parse returned an error: " << err);
+    worker = worker_type{dummy_application{application_results}, ep};
+  }
+
+  bool handle_io_event() override {
+    mpx->handle_updates();
+    return mpx->poll_once(false);
+  }
+
+  multiplexer_ptr mpx;
+  std::shared_ptr<transport_result> transport_results;
+  std::shared_ptr<application_result> application_results;
+  dummy_transport transport;
+  worker_type worker;
+  ip_endpoint ep;
+};
+
+} // namespace
+
+CAF_TEST_FIXTURE_SCOPE(endpoint_manager_tests, fixture)
+
+CAF_TEST(construction and initialization) {
+  CAF_CHECK_EQUAL(worker.init(transport), none);
+  CAF_CHECK_EQUAL(application_results->initialized, true);
+}
+
+CAF_TEST(handle_data) {
+  auto test_span = make_span(reinterpret_cast<byte*>(
```

Comment:
```suggestion
  auto test_span = as_bytes(make_span(hello_test));
```

Reply:
This does not work. `handle_data()` requires a `span<byte>` not a `span<const byte>`. This wont compile with this fix.
Error:
```
invalid conversion from ‘caf::span<const caf::byte>::pointer’ {aka ‘const caf::byte*’} to ‘caf::span<caf::byte>::pointer’ {aka ‘caf::byte*’} [-fpermissive]
```

Reply:
Take a look at the API: there's `as_writable_bytes` for that.

Reply:
I missed that, but still.. I cant cast away `const` using a `reinterpret_cast` I have to use `const_cast<byte*>(ptr)` at some point, or am I missing something else here?

Reply:
Isn't the error that `transport_worker` takes `span<byte>` as argument to `handle_data` instead of `span<const byte>`? To me it looks like giving a mutable buffer to the worker shouldn't happen here, or am I missing something?

Reply:
That was it. I fixed it. Thank you! :+1: